docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,429 @@
+  and the use cases for routing from a gateway and between services are similar
+- Gateway API has different resources per protocol
+
+## Considered Options
+
+- Two resources: `MeshGatewayRoute` and `MeshTrafficRoute`
+- Unified resource for services and gateway: `MeshTrafficRoute`
+- Unified resources but one per protocol, e.g. `MeshHTTPRoute`, `MeshTCPRoute`
+
+## Decision Outcome
+
+The new resource routes and alters requests depending on where it's coming from and where it's going to. Routes end up in the client-side Envoy config so we choose a `spec.to` resource.
+
+```yaml
+spec:
+ targetRef: ... # from where requests are coming
+ to:
+  - targetRef: ... # to where requests are going
+ default: ...
+```
+
+Each protocol gets its own resource:
+
+- `MeshHTTPRoute`
+- `MeshTCPRoute`
+
+which can be extended with others like GRPC.
+
+The route that's used for a given request is the most specific one that Kuma can
+apply given the `kuma.io/protocol` tag.
+
+### Positive Consequences
+
+- Route configuration can be very precisely targeted
+- Users need know only one set of resources for routing
+- A final spec that has the best of `MeshGatewayRoute` and `TrafficRoute`
+
+### New `MeshHTTPRoute` spec
+
+Routing configuration consists of a list of rules where each rule _matches_ some
+requests or traffic, modifies it in some way with _filters_ and then sends it to a _backend_.
+
+#### Matching
+
+The new route matches requests based on:
+
+- headers
+- path
+- method
+- query parameters
+
+##### Headers
+
+Headers can be tested for:
+
+- presence
+- absence
+- exact value
+- prefix value
+- regex value
+
+##### Spec
+
+A rule contains a list of match objects, each of which specifies a match for one
+or more of the above types:
+
+```yaml
+  matches:
+  - method:
+    - GET
+    - POST
+    path:
+      type: RegularExpression|Exact|Prefix
+      value: ...
+    headers:
+      - type: RegularExpression|Exact|Prefix|Absent
+        name: "..."
+        value: ...
+    queryParams:
+      - type: RegularExpression|Exact
+        name: "..."
+        value: ...
+```
+
+A given entry in `match` succeeds if _all_ of the conditions are satisfied and the rule matches when _any one_ of the entries under `match` suceeds. In other words, `match` is an OR'd list of AND'd conditions.
+
+#### Request modification
+
+With `filters` we can change:
+
+- request headers
+- response headers
+- request mirroring (response is ignored)
+- HTTP redirect
+- host
+- request path
+
+```yaml
+  filters:
+  - type: RequestHeaderModifier|ResponseHeaderModifier|RequestMirror|RequestRedirect|URLRewrite
+    requestHeaderModifier:
+      set:
+      - name:
+        value:
+      add:
+      - name:
+        value:
+      remove:
+      - "name"
+    responseHeaderModifier:
+      set:
+      - name:
+        value:
+      add:
+      - name:
+        value:
+      remove:
+      - name
+    requestMirror:
```

Comment:
Gateway API has it as a filter

Reply:
I'd keep it as a filter just to follow gatewayAPI. guidelines filter vs policy is itneresting :)

Reply:
Yes, I understand that Gateway API is a filter. Would be interesting topic anyway (to GAMMA to I think) what should be a filter, what should be a policy.

Reply:
I can't say it's unnatural to do as part of a routing policy since it's really a type of traffic routing. It also fits as part of `filters` IMO since it's a kind of additive modification/handling of a request.

I guess the disadvantage is that for a simple policy of "mirror all my traffic to this destination", it can become cumbersome to make sure that appears as a filter in the right place under all `TrafficRoute` rules.

So maybe it's a question to raise at GAMMA in the context of whether we need route merging

Reply:
Going to close this